test/microbench/org/apache/cassandra/test/microbench/sstable/SSTableReaderBench.java

```diff
@@ -0,0 +1,209 @@
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.UpdateBuilder;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.compaction.OperationType;
+import org.apache.cassandra.db.lifecycle.LifecycleTransaction;
+import org.apache.cassandra.db.rows.UnfilteredRowIterator;
+import org.apache.cassandra.io.sstable.ISSTableScanner;
+import org.apache.cassandra.io.sstable.SSTable;
+import org.apache.cassandra.io.sstable.format.PartitionIndexIterator;
+import org.apache.cassandra.io.sstable.format.SSTableFormat;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.format.SSTableWriter;
+import org.apache.cassandra.io.util.FileUtils;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 1, time = 5)
+@Measurement(iterations = 3, time = 5)
+@Fork(value = 1, jvmArgsAppend = "-Xmx4G")
+@Threads(1)
+@State(Scope.Benchmark)
+public class SSTableReaderBench extends AbstractSSTableBench
+{
+    private final static Logger logger = LoggerFactory.getLogger(SSTableReaderBench.class);
+
+    int KEY_SIZE = 8;
+    int P_KEYS = 10000;
+    int C_KEYS = 1000;
+    int VAL_SIZE = 1;
+
+    public ByteBuffer[] ckeys;
+    public DecoratedKey[] pkeys;
+    public DecoratedKey[] nonpkeys;
+
+    private SSTableReader sstr;
+
+    private int idx = 0;
+
+    private final int step = P_KEYS / 2 - 1;
+
+    @Param({ "table", "table_with_clustering" })
+    public String tableName;
+
+    @Param({ "BIG", "BTI" })
+    public String formatName;
+    private ColumnFamilyStore table;
+    private LifecycleTransaction txn;
+
+    @Setup
+    public void setup()
+    {
+        Keyspace ks = prepareMetadata();
+        table = ks.getColumnFamilyStore(tableName);
+        pkeys = prepareDecoratedKeys(0, P_KEYS, KEY_SIZE);
+        nonpkeys = prepareDecoratedKeys(P_KEYS, P_KEYS * 2, KEY_SIZE);
+        ckeys = prepareBuffers(0, C_KEYS, KEY_SIZE);
+    }
+
+    @Setup(Level.Iteration)
+    public void setupIteration() throws Exception
+    {
+        txn = LifecycleTransaction.offline(OperationType.WRITE);
+        sstr = prepareReader(getFormat(formatName), table, txn);
+    }
+
+    private int nextIdx() {
```

Comment:
maybe it is late... but it is a bit cryptic: what is the invariant for idx?

Reply:
quasi - random